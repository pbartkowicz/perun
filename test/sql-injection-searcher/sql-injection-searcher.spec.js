const fs = require('fs')
const path = require('path')

const SqlInjectionSearcher = require('../../src/sql-injection-searcher')

describe('SqlInjectionSearcher', () => {

    /**
     * @var {SqlInjectionSearcher}
     */
    let searcher

    beforeEach(() => {
        searcher = new SqlInjectionSearcher()
    })

    afterEach(() => {
        jest.restoreAllMocks()
    })

    it('should instantiate', () => {
        expect(() => {
            const searcher = new SqlInjectionSearcher()
            expect(searcher).toBeTruthy()
        }).not.toThrow()
    })

    it('should correctly set sql injection vulnerabilities regexes', () => {
        expect(searcher.regexes.length).toBeGreaterThan(0)
    })

    describe('search', () => {
        const invalidFilesPath = path.resolve(__dirname, './invalid-files')
        const validFilesPath = path.resolve(__dirname, './valid-files')

        const invalidFiles = fs.readdirSync(invalidFilesPath)
        const validFiles = fs.readdirSync(validFilesPath)

        // Number after _ means how many problems should be found in file
        test.each(invalidFiles)('it should find problems in file %s', (filename) => {
            const contents = fs.readFileSync(path.join(invalidFilesPath, filename)).toString()
            const expectedErrorsCount = parseInt(filename.split('_')[1].split('.')[0])

            const results = searcher.search(filename, contents)

            expect(results.valid).toBe(false)
            expect(results.problems).toHaveLength(expectedErrorsCount)
            expect(results).toMatchSnapshot()
        })

        test.each(validFiles)('it should not find problems in file %s', (filename) => {
            const contents = fs.readFileSync(path.join(validFilesPath, filename)).toString()

            const results = searcher.search(filename, contents)

            expect(results.valid).toBe(true)
            expect(results.problems).toHaveLength(0)
        })
    })
})
