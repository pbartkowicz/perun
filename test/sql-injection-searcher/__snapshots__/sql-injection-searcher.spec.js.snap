// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SqlInjectionSearcher search it should find problems in file invalid_1.cpp 1`] = `
Object {
  "problems": Array [
    Object {
      "file": "invalid_1.cpp",
      "line": Object {
        "contents": "SACommand select(&con, _TSA(\\"SELECT NAME, AGE FROM EMPLOYEES WHERE AGE > \\" + age));",
        "number": 2,
      },
      "type": "Potential sql injection vulnerability",
    },
  ],
  "valid": false,
}
`;

exports[`SqlInjectionSearcher search it should find problems in file invalid_2.java 1`] = `
Object {
  "problems": Array [
    Object {
      "file": "invalid_2.java",
      "line": Object {
        "contents": "+ \\"WHERE username = \\" + username)",
        "number": 19,
      },
      "type": "Potential sql injection vulnerability",
    },
    Object {
      "file": "invalid_2.java",
      "line": Object {
        "contents": "+ \\"WHERE user_other_field=\\"+user_other_field)",
        "number": 24,
      },
      "type": "Potential sql injection vulnerability",
    },
  ],
  "valid": false,
}
`;

exports[`SqlInjectionSearcher search it should find problems in file invalid_2.php 1`] = `
Object {
  "problems": Array [
    Object {
      "file": "invalid_2.php",
      "line": Object {
        "contents": "$results = DB::select( DB::raw(\\"SELECT * FROM some_table WHERE some_col = '$someVariable'\\") );",
        "number": 3,
      },
      "type": "Potential sql injection vulnerability",
    },
    Object {
      "file": "invalid_2.php",
      "line": Object {
        "contents": "$results = DB::select( DB::raw(\\"SELECT * FROM some_table WHERE some_other_col = $some_other_variable\\") );",
        "number": 7,
      },
      "type": "Potential sql injection vulnerability",
    },
  ],
  "valid": false,
}
`;

exports[`SqlInjectionSearcher search it should find problems in file invalid_2.sql 1`] = `
Object {
  "problems": Array [
    Object {
      "file": "invalid_2.sql",
      "line": Object {
        "contents": "SELECT * FROM USERS WHERE ID = 15 OR 1=1;",
        "number": 1,
      },
      "type": "Potential sql injection vulnerability",
    },
    Object {
      "file": "invalid_2.sql",
      "line": Object {
        "contents": "SELECT * FROM PRODUCT WHERE ID = 15 UNION SELECT username;",
        "number": 2,
      },
      "type": "Potential sql injection vulnerability",
    },
  ],
  "valid": false,
}
`;

exports[`SqlInjectionSearcher should correctly set sql injection vulnerabilities regexes 1`] = `
Array [
  /OR 1=1/,
  /UNION SELECT username/,
  /WHERE \\[a-zA-Z0-9_\\]\\+ \\?\\[=><\\] \\?" \\?\\[\\+\\.\\] \\?\\\\\\$\\?\\[a-zA-Z0-9_\\]\\+/,
  /WHERE \\[a-zA-Z0-9_\\]\\+ \\?\\[=><\\] \\?'\\?\\\\\\$\\[a-zA-Z0-9_\\]\\+'\\?/,
]
`;
